### ISSUE TITLE
Native Notifications

### ISSUE THREAD

**Issue by hluk** [hluk/CopyQ#1503]
CopyQ uses own implementation of notifications. This has few **drawbacks**:
- It doesn't work well on Wayland (notifications steal keyboard focus - possible issue in Qt with tooltips).
- Obscure native notifications.
- Non-native look and behavior.

**Advantage** of using own implementation:
- No additional library is needed.
- The notifications look and behavior is very similar all platforms.

**Solution** could be to use [KNotifications](https://api.kde.org/frameworks/knotifications/html/index.html). Test code is in branch [knotifications](https://github.com/hluk/CopyQ/commits/knotifications). This has few problems (tested with KNotifications v5.75.0):
- Windows build requires SnoreToast (uses [Microsoft::WRL](https://docs.microsoft.com/en-us/cpp/cppcx/wrl/microsoft-wrl-namespace)) which in turn must be compiled with Visual Studio compiler and would require installing some additional libraries during setup (AFAIK these cannot be included in protable zip file). Alternative is to build KNotifications without SnoreToast but popup notification won't be available and it requires Phonon+DBus support.
- macOS build takes a long time because of installing kde-mac/kde/kf5-knotifications (from homebrew tap https://invent.kde.org/packaging/homebrew-kde).

Related issues:
- #688
- #1247

**Comment by amerlyq** [comment]
Can't we simply use `notify-send(1)` under linux from `libnotify` package?
After all it uses standard API and everyone is able to customize his own notification daemon, whatever their desktop environment provides (or whatever they like). Why hardcode KDE deps?

**Comment by hluk** [comment]
AFAIK, `notify-send` doesn't support buttons in notifications. Example command to create such notification in CopyQ:

```bash
copyq notification .title TITLE .message MESSAGE .button OK "copyq popup 'OK clicked' -" "data 1" .button Close "copyq popup 'Close clicked' -" "data 2"
```

![image](https://user-images.githubusercontent.com/38054/97604373-b34d0b80-1a0d-11eb-83f9-b9ce92d492cb.png)

I believe that in most Linux distributions it wouldn't be a big issue to use KNotifications. It's definitely much simpler to set up than in Windows.

It would be probably useful to set a custom command for notifications anyway. Currently, it's partially possible by overriding `popup()` and `notification()` script functions but sometimes the notifications are created directly without calling the functions.

```ini
[Command]
Command="
    global.popup = function(title, message, time) {
        if (!time)
            time = 4000
    
        if (!message)
            message = ''
    
        execute(
            'notify-send',
            '--app-name=CopyQ',
            '--icon=copyq-normal',
            '--expire-time=' + time,
            title, message)
    }
    
    // This ignores custom buttons.
    global.notification = function() {
        var args = ['notify-send', '--app-name=CopyQ']
        var summary = ''
        var body = ''
        for (var i in arguments) {
            var arg = arguments[i]
            if (arg == '.title') {
                summary = str(arguments[++i])
            } else if (arg == '.message') {
                body = str(arguments[++i])
            } else if (arg == '.time') {
                args.push('--expire-time=' + arguments[++i])
            } else if (arg == '.icon') {
                args.push('--icon=' + arguments[++i])
            }
        }
        args.push(summary)
        args.push(body)
        
        execute.apply(this, args)
    }"
Icon=\xf0a1
IsScript=true
Name=Native Notifications
```

**Comment by resolritter** [comment]
@hluk 

> AFAIK, `notify-send` doesn't support buttons in notifications.

It believe it does.

- [Python's notify2 has an user API for it](https://notify2.readthedocs.io/en/latest/#notify2.Notification.add_action)
- [Implementation: Loop listening on DBus for user interaction with the buttons](https://bitbucket.org/takluyver/pynotify2/src/3ddc2d22632a35bd4c12ebb5684539ce02fecd6c/notify2.py#lines-109)
- [This script](https://github.com/phuhl/notify-send.py#differences-between-notify-sendpy-and-notify-send) allows you to send a notification with buttons from the shell (`--action` option)

Of course, the notification daemon actually has to support the feature. I use one which does ([deadd-notification-center](https://github.com/phuhl/linux_notification_center)). e.g. Ubuntu comes with a different one built-in and I'm not sure if works there.

**Comment by hluk** [comment]
@resolritter That looks like a different script. On my machine:

```
❯ notify-send --version
notify-send 0.7.9

❯ notify-send --help                                                                           
Usage:
  notify-send [OPTION…] <SUMMARY> [BODY] - create a notification

Help Options:
  -?, --help                        Show help options

Application Options:
  -u, --urgency=LEVEL               Specifies the urgency level (low, normal, critical).
  -t, --expire-time=TIME            Specifies the timeout in milliseconds at which to expire the notification.
  -a, --app-name=APP_NAME           Specifies the app name for the icon
  -i, --icon=ICON[,ICON...]         Specifies an icon filename or stock icon to display.
  -c, --category=TYPE[,TYPE...]     Specifies the notification category.
  -h, --hint=TYPE:NAME:VALUE        Specifies basic extra data to pass. Valid types are int, double, string and byte.
  -v, --version                     Version of the package.
```

The executable (not Python script) is provided by libnotify package on Fedora 33.

**Comment by resolritter** [comment]
Oh yes, sorry, I've jumped to conclusions. I thought it was implied that `libnotify` can't do buttons on notifications, which I believe it can... I wanted to clarify that `notify-send` (the program) doesn't have this option and that this limitation is not tied to how the underlying library works. You should be able to have the button interactions using `libnotify` without needing this custom notification runtime, if that was the only showstopper.

The mechanism for those events can be read [here](https://developer.gnome.org/notification-spec/#protocol). And I've showed that `"actions"`, which describes the buttons, is part of the specification. The Arch Wiki page lists a couple of notification servers [here](https://wiki.archlinux.org/index.php/Desktop_notifications#Notification_servers), including the one I've mentioned before ([deadd-notification-center](https://github.com/phuhl/linux_notification_center)) which does have support for buttons. I have not searched if other notification servers also support it but, again, since it's part of the specification, it's more likely to be widespread.

**Comment by Chealer** [comment]
CopyQ 13 uses native notifications, at least on Microsoft Windows.

@hluk: Can you clarify what remains to be solved here?
At least on Windows, one advantage of native notifications is they basically solve issue #1865. But one drawback is that the sound is somewhat excessive for an event as banal as a copy confirmation. As explained in ticket #2335, ClipClip uses 2 distinct sounds, depending on whether the content is already in the clipboard. And both are shorter than the default sound Windows 11 uses. I suggest keeping classic notifications on platforms which do not allow control over the sound.

**Comment by hluk** [comment]
Thanks, Chealer. I think this can be closed. Native notifications are supported and enabled default. The built-in notifications are also still available.
